query-algebrizer/src/lib.rs

```diff
@@ -47,6 +50,10 @@ impl AlgebraicQuery {
                 },
         };
     }
+
+    pub fn cannot_succeed(&self) -> bool {
```

Comment:
Success is a little different than "can return results".  Consider something like `is_empty`, although I suppose that's a property of the `QueryResult` (set) rather than the query itself.  Maybe `query_result_is_empty`?